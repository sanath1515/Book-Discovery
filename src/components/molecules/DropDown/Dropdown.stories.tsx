import React, { ComponentProps } from "react";
import { Story } from "@storybook/react";
import DropDown, { DropDownProps } from "./Dropdown";

const DropDownStory = {
	title: "Molecules/DropDown",
	component: DropDown,
};

const Template: Story<ComponentProps<typeof DropDown>> = (
	args: DropDownProps
) => {
	return <DropDown {...args} />;
};

export const Dropdown = Template.bind({});
Dropdown.args = {
	selectedId: "1",
	items: [
		{
			id: "1",
			name: "Physics",
		},
		{
			id: "2",
			name: "Chemistry",
		},
		{
			id: "3",
			name: "Biology",
		},
	],
};

export const DisabledDropdown = Template.bind({});
DisabledDropdown.args = {
	selectedId: "2",
	items: [
		{
			id: "1",
			name: "None",
		},
		{
			id: "2",
			name: "a-z",
		},
		{
			id: "3",
			name: "z-a",
		},
	],
};

export default DropDownStory;
