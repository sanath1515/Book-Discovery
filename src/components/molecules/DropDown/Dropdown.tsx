import Select from "@material-ui/core/Select";
import React, { useCallback } from "react";
import { makeStyles } from "@material-ui/core/styles";
import { FormControl, MenuItem } from "@material-ui/core";
import theme from "../../../theme/theme";
import { COLORS } from "../../../theme/constants";

const useStyles = makeStyles(() => ({
	formControl: {
		borderRadius: theme.spacing(1),
		border: `solid 1px ${COLORS.GREY_300}`,

		height: "38px",
		justifyContent: "center",
		alignItems: "baseline",
	},
	select: {
		paddingLeft: theme.spacing(3),
		marginRight: `${theme.spacing(1)}px !important`,
		alignItems: "center",
		color: COLORS.CHARCOAL_GREY

	},
	root: {
		alignItems: "center",
		"&:hover": {
			color: "white",
			backgroundColor: "blue",
		},
		

	},
	menuItemSelected: {
		alignItems: "center",
	},
	icon: {
		fontSize: "x-large",
		marginRight: theme.spacing(1),
		height: "20px",
		width: "20px",
	},
	selectRoot: {
		"&:focus": {
			backgroundColor: theme.palette.secondary.light,
		},
	},
}));

export type Item = {
	id: string;
	name: string;
};

export type DropDownProps = {
	selectedId?: string;
	items: Array<Item>;
	onChange: (
		event: React.ChangeEvent<{ name?: string; value: unknown }>
	) => void;
	testId?: string;
	defaultValue?: string;
	placeHolderText?: string;
	ifEmptyMessage?: string;
	valueSelected?: string;
};

const DropDown: React.FC<DropDownProps> = ({
	selectedId,
	onChange,
	items,
	testId,
	defaultValue,
	placeHolderText,
	ifEmptyMessage,
	valueSelected,
}) => {
	const classes = useStyles();

	const getRenderValue = useCallback(
		(id?: string, placeholderText?: string, valueSelectedOf?: string) => {
			if (valueSelectedOf === "NONE") {

				return () => placeHolderText;
			}
			return id ? undefined : () => placeholderText;
		},
		[]
	);

	return (
		<FormControl className={classes.formControl}>
			<Select
				value={selectedId}
				onChange={onChange}
				displayEmpty={defaultValue ? false : true}
				disableUnderline
				classes={{
					select: classes.select,
					icon: classes.icon,
					// root: classes.selectRoot,

				}}
				data-testid="option"
				MenuProps={{ disableScrollLock: true }}
				defaultValue={defaultValue}
				renderValue={
					placeHolderText
						? getRenderValue(selectedId, placeHolderText, valueSelected)
						: undefined
				}
			>
				{!items.length && ifEmptyMessage && (
					<MenuItem
						key={ifEmptyMessage}
						value=""
						classes={{
							root: classes.root,
						}}
						data-testid="empty_value"
					>
						{ifEmptyMessage}
					</MenuItem>
				)}
				{items.map((item: Item, index) => {
					return (
						<MenuItem
							key={item.id}
							value={item.id}
							classes={{
								root: classes.root,
							}}
						>
							{item.name}
						</MenuItem>
					);
				})}
			</Select>
		</FormControl>
	);
};

export default DropDown;
