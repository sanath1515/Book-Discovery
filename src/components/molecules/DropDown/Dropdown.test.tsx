import { render, fireEvent, screen } from "@testing-library/react";
import DropDown from "./Dropdown";
import { act } from "react-dom/test-utils";

describe("Dropdown test", () => {
	const handleChange = jest.fn();

	it("should render dropdown correctly", () => {
		const { queryByText } = render(
			<DropDown
				selectedId="1"
				items={[
					{ id: "1", name: "value1" },
					{ id: "2", name: "value2" },
				]}
				onChange={handleChange}
			/>
		);
		expect(queryByText("value1")).toBeTruthy();
	});

	it("should match snapshot", () => {
		const { queryByText } = render(
			<DropDown
				selectedId="1"
				items={[
					{ id: "1", name: "value1" },
					{ id: "2", name: "value2" },
				]}
				onChange={handleChange}
			/>
		);
		expect(queryByText("value1")).toMatchSnapshot();
	});
	it("should call handle change when dropdown option is changed", () => {
		const { container, queryByText, getAllByTestId } = render(
			<DropDown
				selectedId="1"
				placeHolderText="Sort by"
				items={[
					{ id: "1", name: "value1" },
					{ id: "2", name: "value2" },
				]}
				onChange={handleChange}
			/>
		);

		act(() => {
			const options = getAllByTestId("option");
			fireEvent.mouseDown(options[0]);
			options[0].click();
		});
		expect(handleChange).toHaveBeenCalled();

		fireEvent.change(screen.getByTestId("option"), {
			target: { value: "NONE" },
		});
		expect(screen.getByText("Sort by")).toBeInTheDocument();
		expect(handleChange).toHaveBeenCalled();
	});
});
