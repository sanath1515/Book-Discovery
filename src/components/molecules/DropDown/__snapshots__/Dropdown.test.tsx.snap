// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown test should match snapshot 1`] = `
<div
  aria-haspopup="listbox"
  class="MuiSelect-root MuiSelect-select makeStyles-select-8 MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
  role="button"
  tabindex="0"
>
  value1
</div>
`;
